pkg/gadgets/trace/dns/tracer/tracer.go

```diff
@@ -332,30 +132,80 @@ func (t *Tracer) install() error {
 	return nil
 }
 
-func (t *Tracer) run(ctx context.Context, logger logger.Logger, dnsTimeout time.Duration) error {
-	spec, err := loadDns()
+func (t *Tracer) parseDNSPacket(rawSample []byte, netns uint64) (*types.Event, error) {
+	// The sample received is a concatenation of the dnsEventT structure and the packet bytes.
+	bpfEvent := (*dnsEventT)(unsafe.Pointer(&rawSample[0]))
+	packetBytes := rawSample[unsafe.Sizeof(*bpfEvent):]
+
+	// The DNS layer starts after the ethernet, IP and UDP headers.
+	DNS_OFF := 14 + 20 + 8
+	dnsLayer := layers.DNS{}
+	err := dnsLayer.DecodeFromBytes(packetBytes[DNS_OFF:], gopacket.NilDecodeFeedback)
 	if err != nil {
-		return fmt.Errorf("loading asset: %w", err)
+		return nil, fmt.Errorf("decoding dns layer: %w", err)
 	}
 
-	parseDNSEvent := func(rawSample []byte, netns uint64) (*types.Event, error) {
-		bpfEvent := (*dnsEventTAbbrev)(unsafe.Pointer(&rawSample[0]))
-		// 4 (padding) + (size of event without addresses) + (address count * 16)
-		expected := 4 + int(unsafe.Sizeof(*bpfEvent)) + int(bpfEvent.Anaddrcount)*16
-		if len(rawSample) != expected {
-			return nil, fmt.Errorf("invalid sample size: received: %d vs expected: %d",
-				len(rawSample), expected)
-		}
+	ipversion := gadgets.IPVerFromAF(bpfEvent.Af)
 
-		event, err := bpfEventToDNSEvent(bpfEvent, rawSample[unsafe.Offsetof(dnsEventT{}.Anaddr):], netns)
-		if err != nil {
-			return nil, err
+	event := &types.Event{
+		Event: eventtypes.Event{
+			Type:      eventtypes.NORMAL,
+			Timestamp: gadgets.WallTimeFromBootTime(bpfEvent.Timestamp),
+		},
+		WithNetNsID:   eventtypes.WithNetNsID{NetNsID: netns},
+		WithMountNsID: eventtypes.WithMountNsID{MountNsID: bpfEvent.MountNsId},
+		Pid:           bpfEvent.Pid,
+		Tid:           bpfEvent.Tid,
+		Uid:           bpfEvent.Uid,
+		Gid:           bpfEvent.Gid,
+		Comm:          gadgets.FromCString(bpfEvent.Task[:]),
+		PktType:       pktTypeToString(bpfEvent.PktType),
+
+		SrcIP:    gadgets.IPStringFromBytes(bpfEvent.SaddrV6, ipversion),
+		DstIP:    gadgets.IPStringFromBytes(bpfEvent.DaddrV6, ipversion),
+		SrcPort:  bpfEvent.Sport,
+		DstPort:  bpfEvent.Dport,
+		Protocol: gadgets.ProtoString(int(bpfEvent.Proto)),
+
+		ID:         fmt.Sprintf("%.4x", dnsLayer.ID),
+		NumAnswers: int(dnsLayer.ANCount),
+	}
+
+	if dnsLayer.QR {
+		event.Qr = types.DNSPktTypeResponse
+		event.Nameserver = event.SrcIP
+
+		event.Latency = time.Duration(bpfEvent.LatencyNs)
+		event.Rcode = rcodeToString(uint8(dnsLayer.ResponseCode))
```

Comment:
```suggestion
		event.Rcode = dnsLayer.ResponseCode.String()
```
?

Reply:
The string definitions of both places aren't exactly the same:

```golang
// DNS header RCODE (response code) field.
// https://datatracker.ietf.org/doc/rfc1035#section-4.1.1
func rcodeToString(rcode uint8) string {
	switch rcode {
	case 0:
		return "NoError"
	case 1:
		return "FormErr"
	case 2:
		return "ServFail"
	case 3:
		return "NXDomain"
	case 4:
		return "NotImp"
	case 5:
		return "Refused"
	default:
		return fmt.Sprintf("Unknown(%d)", rcode)
	}
}
```

vs 

```golang
func (drc DNSResponseCode) String() string {
	switch drc {
	default:
		return "Unknown"
	case DNSResponseCodeNoErr:
		return "No Error"
	case DNSResponseCodeFormErr:
		return "Format Error"
	case DNSResponseCodeServFail:
		return "Server Failure "
	case DNSResponseCodeNXDomain:
		return "Non-Existent Domain"
	case DNSResponseCodeNotImp:
		return "Not Implemented"
	case DNSResponseCodeRefused:
		return "Query Refused"
	case DNSResponseCodeYXDomain:
		return "Name Exists when it should not"
	case DNSResponseCodeYXRRSet:
		return "RR Set Exists when it should not"
	case DNSResponseCodeNXRRSet:
		return "RR Set that should exist does not"
	case DNSResponseCodeNotAuth:
		return "Server Not Authoritative for zone"
	case DNSResponseCodeNotZone:
		return "Name not contained in zone"
	case DNSResponseCodeBadVers:
		return "Bad OPT Version"
	case DNSResponseCodeBadKey:
		return "Key not recognized"
	case DNSResponseCodeBadTime:
		return "Signature out of time window"
	case DNSResponseCodeBadMode:
		return "Bad TKEY Mode"
	case DNSResponseCodeBadName:
		return "Duplicate key name"
	case DNSResponseCodeBadAlg:
		return "Algorithm not supported"
	case DNSResponseCodeBadTruc:
		return "Bad Truncation"
	case DNSResponseCodeBadCookie:
		return "Bad Cookie"
	}
}
```

I don't want to break users relying on this.

Reply:
`gopacket` is a well-known pkg, so we can consider it more "standard". I think that's a good justification to migrate and use it. We can alert users in the release notes and Slack about this breaking change.